sdk/program/src/message/legacy.rs

```diff
@@ -573,12 +579,34 @@ impl Message {
     /// instructions in this message. Since the dynamic set of reserved accounts
     /// isn't used here to demote write locks, this shouldn't be used in the
     /// runtime.
-    pub fn is_maybe_writable(&self, i: usize) -> bool {
+    pub fn is_maybe_writable(
+        &self,
+        i: usize,
+        reserved_account_keys: Option<&HashSet<Pubkey>>,
+    ) -> bool {
         (self.is_writable_index(i))
-            && !is_builtin_key_or_sysvar(&self.account_keys[i])
+            && !self.is_account_maybe_reserved(i, reserved_account_keys)
             && !self.demote_program_id(i)
     }
 
+    /// Returns true if the account at the specified index is in the reserved
+    /// account keys set. Before loading addresses, we can't detect reserved
+    /// account keys properly so this shouldn't be used by the runtime.
+    fn is_account_maybe_reserved(
+        &self,
+        key_index: usize,
+        reserved_account_keys: Option<&HashSet<Pubkey>>,
+    ) -> bool {
+        if let Some(reserved_account_keys) = reserved_account_keys {
+            self.account_keys
+                .get(key_index)
+                .map(|key| reserved_account_keys.contains(key))
+                .unwrap_or_default()
```

Comment:
Just to make sure, this changes behavior because it won't panic if `key_index` is out of range. Since this is only used in transaction-status and cli-output, which are doing things right, there isn't too much concern.

Others might use this incorrectly and be easily confused by the results. How about changing this and `is_maybe_writable` to return a `Result` instead? Or create `try_is_maybe_writable` to return the `Result` and have `is_maybe_writable` unwrap it and panic?

Reply:
Ah right, thanks for pointing that out, all good then!